backend/src/routes/user.rs

```diff
@@ -0,0 +1,275 @@
+            email: Set(Some(user.email.to_string())),
+            salt: Set(Some(salt.to_string())),
+            ..Default::default()
+        };
+        // insert the row in database
+        let ins_result = users.insert(&db).await;
+        match ins_result {
+            Ok(res) => {
+                info!("User signup Succ, save user: {:?}", res.uid);
+                (
+                    Status::Ok,
+                    Some(Json(UserResponse {
+                        errors: error_collector,
+                    })),
+                )
+            }
+            _ => (Status::InternalServerError, None),
+        }
+    }
+}
+
+#[post("/login", data = "<user_info>", format = "json")]
+pub async fn user_log_in(
+    db: Connection<PgDb>,
+    kvdb: Connection<RedisDb>,
+    cookies: &CookieJar<'_>,
+    user_info: Json<UserLoginInfo<'_>>,
+) -> (Status, Option<String>) {
+    let mut con = kvdb.into_inner();
+    // get user info from request
+    let user = user_info.into_inner();
+    // check if username is existed, add corresponding error if so
+    let username_existence_result = User::find()
+        .filter(pgdb::user::Column::Username.eq(user.username))
+        .one(&db)
+        .await;
+    // check if password is wrong, add corresponding error if so
+    match username_existence_result {
+        Ok(s) => match s {
+            Some(matched_user) => {
+                info!("username exists, continue...");
+                let salt = match matched_user.salt {
+                    Some(s) => s,
+                    None => {
+                        error!("cannot find user's salt.");
+                        return (Status::InternalServerError, None);
+                    }
+                };
+                // encrypt input password same as sign-up
+                let mut hash_sha3 = Sha3::sha3_256();
+                hash_sha3.input_str(&(String::from(&salt) + user.password));
+                let password = hash_sha3.result_str();
+                if matched_user.password.eq(&Some(password.to_string())) {
+                    info!("password correct, continue...");
+                    // generate token and refresh token
+                    let token: String = iter::repeat(())
+                        .map(|()| thread_rng().sample(Alphanumeric))
+                        .map(char::from)
+                        .take(32)
+                        .collect();
+                    let mut hash_sha3 = Sha3::sha3_384();
+                    hash_sha3.input_str(&token);
+                    let refresh_token = hash_sha3.result_str();
+                    // build cookie
+                    let cookie = Cookie::build("token", token.clone())
+                        .domain("thuburrow.com")
+                        .path("/")
+                        .same_site(SameSite::None)
+                        .finish();
+                    // set cookie
+                    cookies.add_private(cookie);
```

Comment:
You should set it after everything is done of redis. Because if you have error in the way, you will early return and at this time, cookie should not be set. Just move this part down to maybe its original part.